sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala

```diff
@@ -302,10 +302,22 @@ class V2SessionCatalog(catalog: SessionCatalog)
 
     val finalProperties = CatalogV2Util.applyClusterByChanges(properties, schema, changes)
     try {
-      catalog.alterTable(
-        catalogTable.copy(
-          properties = finalProperties, schema = schema, owner = owner, comment = comment,
-          collation = collation, storage = storage))
+      if (SQLConf.get.getConf(StaticSQLConf.CATALOG_IMPLEMENTATION).equals("hive")) {
```

Comment:
Note: SessionCatalog (V1) alterTableDataSchema explicitly does not support 'drop column'.  https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala#L499 

So to avoid regression in InMemoryCatalog, in which alterTable can alter the schema, I make the change only for "hive".  Following the config check for CATALOG_IMPLEMENTATION in other parts of the code.

Reply:
Does it break any test? `InMemoryCatalog` is for testing and I think it's OK to always call v1 `alterTableDataSchema` for table schema changes.